File: 0756-pyramid-transition-matrix/0756-pyramid-transition-matrix.cpp
class Solution {
public:
    bool pyramidTransition(string bottom, vector<string>& allowed) {
        unordered_map<string, vector<char>> pattern;
        unordered_set<string> visited;
        for(int i = 0; i < allowed.size(); i++)
        {
            pattern[allowed[i].substr(0, 2)].push_back(allowed[i][2]);
        }
        return solve(bottom, "", pattern, visited, 0);
    }
private:
    bool solve(string bottom, string top, unordered_map<string, vector<char>>& pattern, unordered_set<string>& visited, int i)
    {
        if(bottom.length() == 1)
        {
            return true;
        }
        if(visited.find(top) != visited.end())
        {
            return false;
        }
        if(i == bottom.length() - 1)
        {
            return solve(top, "", pattern, visited, 0);
        }
        string block = bottom.substr(i, 2);
        if(pattern.find(block) != pattern.end())
        {
            for(int idx = 0; idx < pattern[block].size(); idx++)
            {
                if(solve(bottom, top + pattern[block][idx], pattern, visited, i + 1))
                {
                    return true;
                }
            }
        }
        visited.insert(bottom);
        return false;
    }
};

File: 0756-pyramid-transition-matrix/0756-pyramid-transition-matrix.py
def solve(bottom: str, top: str, pattern: dict[str, List[str]], visited: set[str], i: int) -> bool:
    print(top)
    if len(bottom) == 1:
        return True
    if top in visited:
        return False
    if i == len(bottom) - 1:
        return solve(top, '', pattern, visited, 0)
    block = bottom[i:i + 2]
    if block in pattern:
        for c in pattern[block]:
            if solve(bottom, top + c, pattern, visited, i + 1):
                return True
    visited.add(bottom)
    return False
class Solution:
    def pyramidTransition(self, bottom: str, allowed: List[str]) -> bool:
        pattern = {}
        visited = set()
        for a in allowed:
            if a[:2] not in pattern:
                pattern[a[:2]] = []
            pattern[a[:2]].append(a[2])
        return solve(bottom, '', pattern, visited, 0)